source/Various/Hydra.lagda

```diff
@@ -0,0 +1,182 @@
+
+⊲-is-well-founded : is-well-founded _⊲_
+⊲'-is-well-founded : is-well-founded _⊲'_
+
+⊲-is-well-founded (Node hs) = acc (recurs (⊲'-is-well-founded hs))
+ where
+  recurs : {hs : List Hydra}
+         → is-accessible (_⊲'_) hs
+         → (h' : Hydra) → h' ⊲ Node hs
+         → is-accessible (_⊲_) h'
+  recurs (acc rec₁) (Node hs') r = acc (recurs (rec₁ hs' (⊲⇒⊲' r)))
+
+⊲'-is-well-founded [] = acc (λ h r → 𝟘-elim ([]-is-minimum h r))
+⊲'-is-well-founded (h ∷ hs) = acc (recurs (⊲-is-well-founded h) (⊲'-is-well-founded hs))
+ where
+  recurs : {h : Hydra} {hs : List Hydra}
+         → is-accessible _⊲_ h
+         → is-accessible _⊲'_ hs
+         → (hs' : List Hydra) → hs' ⊲' (h ∷ hs)
+         → is-accessible _⊲'_ hs'
+  recurs (acc rec₁) (acc rec₂) hs' (step₀ n here refl)      = acc rec₂
+  recurs (acc rec₁) (acc rec₂) hs' (step₀ n (next l) refl)  = acc (recurs (acc rec₁) (rec₂ _ (step₀ n l refl)))
+  recurs (acc rec₁) (acc rec₂) hs' (step₁ n (next l) refl)  = acc (recurs (acc rec₁) (rec₂ _ (step₁ n l refl)))
+  recurs (acc rec₁) (acc rec₂) hs' (step₁ n (here₁ l) refl) = acc (recurs (rec₁ _ (step n (inr l) refl)) (acc rec₂))
+  recurs (acc rec₁) (acc rec₂) hs' (step₁ n (here₀ l) refl) = recurs' n
+   where
+    recurs' : (n : ℕ) → is-accessible _⊲'_ (cons-mult _ (succ n) _)
+    recurs' 0        = acc (recurs (rec₁ _ (step 0 (inl l) refl)) (acc rec₂))
+    recurs' (succ n) = acc (recurs (rec₁ _ (step 0 (inl l) refl)) (recurs' n))
+
+\end{code}
+
+We can now define Hydra(n) and compute its four first terms.
+
+\begin{code}
+
+leftmost-head₀ : (hs : List Hydra) → hs ≠ [] → HeadLocation₀ hs + HeadLocation₁ hs
+leftmost-head₀ []                    neq = 𝟘-elim (neq refl)
+leftmost-head₀ (Head ∷ _)            neq = inl here
+leftmost-head₀ ((Node (h ∷ hs)) ∷ _) neq = leftmost-head₀' (leftmost-head₀ (h ∷ hs) (λ ()))
+ where
+  leftmost-head₀' : HeadLocation₀ (h ∷ hs) + HeadLocation₁ (h ∷ hs)
+                  → HeadLocation₀ _ + HeadLocation₁ _
+  leftmost-head₀' (inl l) = inr (here₀ l)
+  leftmost-head₀' (inr l) = inr (here₁ l)
+
+leftmost-head : (h : Hydra) → h ≠ Head → HeadLocation h
+leftmost-head Head            neq = 𝟘-elim (neq refl)
+leftmost-head (Node (h ∷ hs)) neq = leftmost-head₀ (h ∷ hs) (λ ())
+
+f-Hydra : (n : ℕ) → ℕ
+f-Hydra n = battle 1 (tall-hydra n) (⊲-is-well-founded _)
+ where
+  tall-hydra : (n : ℕ) → Hydra
+  tall-hydra 0 = Node []
+  tall-hydra (succ n) = Node (tall-hydra n ∷ [])
+
+  battle : ℕ → (h : Hydra) → is-accessible _⊲_ h → ℕ
+  battle turn Head            (acc rec₁ ) = 0
+  battle turn (Node (h ∷ hs)) (acc rec₁ ) = succ (battle (succ turn) (cut turn (Node (h ∷ hs)) cut-head) (rec₁ _ (turn , (cut-head , refl))))
+   where
+    cut-head : HeadLocation (Node (h ∷ hs))
+    cut-head = leftmost-head _ (λ ())
+
+f-Hydra0 : f-Hydra 0 ＝ 0
+f-Hydra0 = refl
+
+f-Hydra1 : f-Hydra 1 ＝ 1
+f-Hydra1 = refl
+
+f-Hydra2 : f-Hydra 2 ＝ 3
+f-Hydra2 = refl
+
+f-Hydra3 : f-Hydra 3 ＝ 37
+f-Hydra3 = refl
+
+\end{code}
```

Comment:
Please add some links to the Hydra game at the beginning of the file. For example,
[1] https://en.wikipedia.org/wiki/Hydra_game

[2] Kirby, Laurie; Paris, Jeff. Bull. Accessible independence results for Peano arithmetic.
     London Math. Soc, 14 (1982), 285-293.
     http://logic.amu.edu.pl/images/3/3c/Kirbyparis.pdf

Also please add a sentence at the very beginning of the file, after your name, describing what this file does.

Thanks for your contribution!